### DISCUSSION TITLE
Custom shapes: Are they documented? What ways do we have to create new node shapes?

### DISCUSSION THREAD

**Discussion by tony** [xyflow/xyflow#1040] [Q&A]
Hi @wbkd!  Thank you for the project!

Question: Is there any documentation on custom shapes for nodes (other than rounded boxes?)

Is there a place to add a custom SVG?  Can there be a doc page on various approaches for customizing nodes if there's multiple?

**Comment by moklick** [comment]
Hey @tony 

React Flow doesn't include any shapes. I would always recommend to create custom nodes. The predefined ones are more like examples. Since React Flow is DOM based you can style your nodes with regular CSS. 

Docs for custom nodes: https://reactflow.dev/docs/api/node-types/ ([example](https://reactflow.dev/examples/custom-node/))
Docs for theming: https://reactflow.dev/docs/theming/

**Reply by tony** [reply]
@moklick Thank you

An example of a different in the docs would be very welcome!

Aside: Congratulations on the HN post! https://news.ycombinator.com/item?id=26860381

**Reply by oohogwash** [reply]
Do all nodes have to be rectangles ?, for example diamond, triangle, circle shapes for the boundary and other more advanced detail inside the boundary of the node. I am not sure if this can be done in what has been documented.

**Reply by tony** [reply]
@oohogwash That's what I feel as well.

@moklick There are many  examples and the documentation is excellent, however, everything (to my knowledge) is the same shape. It's not clear how we could add our own geometries, svg, etc.  

Even more complex: If I want to add circles or have the shape animate / morph (like a game engine), is something like that possible? How much customization is possible?

**Reply by CamboimGabriel** [reply]
@tony 
Its all possible, check how I'm using the nodes
All with different shapes

![image](https://user-images.githubusercontent.com/36650520/116019511-b9b5ec00-a61a-11eb-8260-5368016a06f9.png)

**Reply by tony** [reply]
@CamboimGabriel Can you make a repo/gist for this? Or perhaps contribute to the docs of this projects?

Taken even further: could this get interactive? Could these nodes have a "split" animation after a time/or event like clicking?

Is it possible to make a circle node that splits into a second node like during cell division?  
![cell split](https://user-images.githubusercontent.com/26336/116027857-87f15500-a61b-11eb-9ea1-83c7b384c19d.gif)

(except the cell wouldn't leave the screen, it'd be an item)

**Reply by oohogwash** [reply]
@CamboimGabriel  Are there code examples of how to do what you showed in here. that would be great. @tony  would like to see as well I am sure.

**Reply by oohogwash** [reply]
@CamboimGabriel   I would think that the use of graphics like your demo would have the ability to have tool tips or text aligned to the graphical symbol. If that is part of the demo and code that would be great too.

**Reply by CamboimGabriel** [reply]
I'm using styled components for this, theres no secret on how to implement custom shapes

Here is a code example:

```
<ConditionalEndStyle status={data.status} selected={selected}>
      <ConditionalEndStyleInside>
        <Plus style={{ transform: "rotate(45deg)" }} size={50} />
      </ConditionalEndStyleInside>

      <Handle
        id="b"
        type="target"
        position={targetPosition}
        style={{ top: 0, width: "8px", height: "8px" }}
        isValidConnection={(e) =>
          nodes.find((item) => item.id === e.source).type === "parallel" ||
          nodes.find((item) => item.id === e.source).type === "conditional"
            ? true
            : !edges.find((item) => item.source === e.source)
        }
      />

      <Handle
        type="source"
        position={sourcePosition}
        id="a"
        style={{
          top: "100%",
          width: "8px",
          height: "8px",
        }}
        isValidConnection={(e) =>
          nodes.find((item) => item.id === e.target).type === "parallelEnd" ||
          nodes.find((item) => item.id === e.target).type === "conditionalEnd"
            ? !edges.find((item) => item.source === e.source)
            : !edges.find(
                (item) => item.source === e.source || item.target === e.target
              )
        }
      />
    </ConditionalEndStyle>
```

And here is the styled component of the shape:

```
export const ConditionalEndStyle = styled.div`
  width: 50px;
  height: 50px;
  box-shadow: ${({ selected }) =>
    selected === true ? "0 0 0 1px blue" : " 0 0 0 0.5px #1a192b"};
  box-sizing: border-box;
  border-radius: 8px;
  transform: rotate(-45deg);
  background-color: ${({ status }) =>
    status === "done" ? "#beefbe" : status === "doing" ? "#f2f0a6" : "white"};
  display: flex;
  justify-content: center;
  align-items: center;
`;

```
Here is the final result for that shape: 

![image](https://user-images.githubusercontent.com/36650520/116067523-95303300-a65f-11eb-9c2c-5cf419e3255c.png)

I didnt tried make animations yet, but I think its all possible with CSS

**Reply by moklick** [reply]
I would be happy if someone would do a PR for a shapes example in the [docs repo](https://github.com/wbkd/react-flow-docs). A split animation would be complicated to implement but I am sure it's possible somehow ;)

**Reply by oohogwash** [reply]
@moklick  I would be happy to do PR, however I have not written any code with this stack so it would be poor quality. I would like to be a collaborator to see it evolve if thats possible.

**Reply by oohogwash** [reply]
@CamboimGabriel  i did some playing with css, I can see how the diamond is made and researched items for other shapes. 
A question raises in how the hit test algorithms work with these, will they just work and know what the boundary is ?

**Reply by marcel-burda** [reply]
Hi guys, great example. Is there any chance you have a similiar example with icons as nodes to share :) ?

**Reply by datoslabs** [reply]
@marcel-burda, the example @CamboimGabriel shared is "icons as nodes" (ie `Plus` is an icon in a custom node).

**Reply by CamboimGabriel** [reply]
Here I have some examples with icons (like the timer node)
![image](https://user-images.githubusercontent.com/36650520/138002081-c046f5b7-6390-4bf4-9e80-cc8db6d075f5.png)

**Reply by GadgetGeekUK** [reply]
@CamboimGabriel Apologies for asking again, do you have a full demo for custom shapes? The snippets don't make sense to me yet! Thanks

**Reply by tony** [reply]
I'm in a similar position.  Until examples are there and maintained in the docs, this matter isn't addressed and feels as if its not supported (when it may in fact be)

**Reply by visualsitemaps** [reply]
@GadgetGeekUK can you post your demo on https://codesandbox.io/ or similar?
thanks

**Reply by Krithika3** [reply]
Are there any good examples here? Thanks once again
